File: Java/Search-for-a-Range.java
public class Solution {
    /**
     *@param A : an integer sorted array
     *@param target :  an integer to be inserted
     *return : a list of length 2, [index1, index2]
     */
    public int findleft(ArrayList<Integer> A, int target) {
        if (A == null || A.size() <= 0) return -1;
        int left = 0, right = A.size() - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (A.get(mid) >= target) {
                right = mid;
            } else if (A.get(mid) < target) left = mid;
        }
        if (A.get(left) == target) return left;
        else if (A.get(right) == target) return right;
        return -1;
    }
    public int findRight(ArrayList<Integer> A, int target) {
        if (A == null || A.size() <= 0) return -1;
        int left = 0, right = A.size() - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (A.get(mid) > target) {
                right = mid;
            } else if (A.get(mid) <= target) left = mid;
        }
        if (A.get(right) == target) return right;
        else if (A.get(left) == target) return left;
        return -1;
    }
    public ArrayList<Integer> searchRange(ArrayList<Integer> A, int target) {
        // write your code here
        ArrayList<Integer> res = new ArrayList<Integer>();
        res.add(findleft(A, target));
        res.add(findRight(A, target));
        return res;
    }
}

File: Cpp/Search-for-a-Range.cpp
class Solution {
    /** 
     *@param A : an integer sorted array
     *@param target :  an integer to be inserted
     *return : a list of length 2, [index1, index2]
     */
public:
    vector<int> searchRange(vector<int> &A, int target) {
        // write your code here
        vector<int> res(2);
        res[0] = res[1] = -1;
        if (A.empty()) {
            return res;
        }
        int start = 0, end = A.size() - 1, mid;
        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            if (A[mid] == target) end = mid;
            else if (A[mid] < target) start = mid;
            else end = mid;
        }
        if (A[start] == target) res[0] = start;
        else if (A[end] == target) res[0] = end;
        else return res;
        start = 0, end = A.size() - 1;
        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            if (A[mid] == target) start = mid;
            else if (A[mid] < target) start = mid;
            else end = mid;
        }
        if (A[end] == target) res[1] = end;
        else if (A[start] == target) res[1] = start;
        else return res;
        return res;
    }
};
